pgjdbc/src/test/java/org/postgresql/jdbc/ResourceLockTest.java

```diff
@@ -0,0 +1,139 @@
+    CallWithResourceLock callWithResourceLock = new CallWithResourceLock();
+
+    int levelOfConcurrency = 5;
+
+    ExecutorService executorService = Executors.newFixedThreadPool(levelOfConcurrency);
+    try {
+      List<Callable<CounterPair>> callables = new ArrayList<>();
+      for (int i = 0; i < levelOfConcurrency; i++) {
+        callables.add(callWithResourceLock);
+      }
+
+      // expect linear execution
+      List<Future<CounterPair>> results = executorService.invokeAll(callables);
+
+      List<CounterPair> pairs = new ArrayList<>();
+      Set<Integer> preLockSet = new HashSet<>();
+      Set<Integer> postLockSet = new HashSet<>();
+      for (Future<CounterPair> result : results) {
+        CounterPair entry = result.get();
+        preLockSet.add(entry.preLock);
+        postLockSet.add(entry.postLock);
+        pairs.add(entry);
+      }
+
+      // confirm the callables were executed serially without any overlap
+      Collections.sort(pairs);
+      for (int i = 0; i < pairs.size() - 1; i++) {
+        CounterPair pair = pairs.get(i);
+        CounterPair nextPair = pairs.get(i + 1);
+        assertTrue(pair.beforeWithoutOverlap(nextPair), "executed serially without overlap");
+      }
+
+      assertEquals(1, preLockSet.size(),
+          "preLockSet.size() should equal 1 with only one thread winning the initial race to obtain the lock");
+      assertEquals(levelOfConcurrency, postLockSet.size(),
+          "postLockSet.size() should equal levelOfConcurrency with all callables completed");
+
+    } finally {
+      executorService.shutdown();
+    }
+  }
+
+  static final class CallWithResourceLock implements Callable<CounterPair> {
+
+    // wait enough time to allow concurrent threads to block on the lock
+    final long waitTime = TimeUnit.MILLISECONDS.toNanos(20);
+    final ResourceLock lock = new ResourceLock();
+    final AtomicInteger counter = new AtomicInteger();
+
+    /**
+     * Invoke returning 'pre lock' and 'post lock' counter value.
+     */
+    public CounterPair call() {
+      int preLock = counter.get();
+      try (ResourceLock ignore = lock.obtain()) {
+        long startNanos = System.nanoTime();
+        int postLock = counter.get();
+        LockSupport.parkNanos(waitTime);
+        counter.incrementAndGet();
+        return new CounterPair(startNanos, preLock, postLock);
+      }
+    }
+  }
+
+  static final class CounterPair implements Comparable<CounterPair> {
+    final int preLock;
+    final int postLock;
+    final long startExecutionNanos;
+    final long endExecutionNanos = System.nanoTime();
+
+    CounterPair(long startExecutionNanos, int preLock, int postLock) {
+      this.startExecutionNanos = startExecutionNanos;
+      this.preLock = preLock;
+      this.postLock = postLock;
+    }
+
+    boolean beforeWithoutOverlap(CounterPair other) {
+      return endExecutionNanos < other.startExecutionNanos;
```

Comment:
I'm afraid this use of `System.nanoTime` is not justified:
1) There's a possibility for overflow, so we should not compare absolute values of `System.nanoTime` (see javadoc for `nanoTime`)
2) There's no guarantee on nanosecond resolution (see javadoc), so `nanoTime` can return the same values for different calls. In other words, if `nanoTime` returns `0` always, it would still meet the contract for `nanoTime`.

Reply:
> There's a possibility for overflow

I don't see that as long as we are not comparing values that span over 262 years we should be good as per javadoc (and this is sub 1 second). We could compute the diff between the values and confirm that is greater than zero ... but we are doing the equivalent just without doing the subtraction - I don't see an overflow issue here but happy if you want to adjust this as you see fit.

> There's no guarantee on nanosecond resolution

True, but it is at least the precision of currentTimeMillis and with the task taking 20ms we don't have an issue with precision here. As I see it, historically we could hit an issue on some JDKs sub 10ms but we are above that precision requirement so I see no issue there.


> System.nanoTime is not justified

You are welcome to adjust this how you like. I have no deep need to keep this test as it is.

Reply:
> There's a possibility for overflow

Ah, the use of < has the potential here to not work due to overflow - got it, changed to use System.currentTimeMillis().